File: homework4/design_pattern/csharp/builder.cs
using System;

public class Person
{
    public string Name { get; set; }
    public int Age { get; set; }
    public string Address { get; set; }
}

public class PersonBuilder
{
    private readonly Person _person;

    public PersonBuilder()
    {
        _person = new Person();
    }

    public PersonBuilder WithName(string name)
    {
        _person.Name = name;
        return this; // Return itself for method chaining
    }

    public PersonBuilder WithAge(int age)
    {
        _person.Age = age;
        return this; // Return itself for method chaining
    }

    public PersonBuilder WithAddress(string address)
    {
        _person.Address = address;
        return this; // Return itself for method chaining
    }

    public Person Build()
    {
        return _person;
    }
}

class Program
{
    static void Main(string[] args)
    {
        Person person = new PersonBuilder()
            .WithName("홍길동")
            .WithAge(20)
            .WithAddress("양산시 부산대학로")
            .Build();

        Console.WriteLine($"Name: {person.Name}, Age: {person.Age}, Address: {person.Address}");
    }
}

File: homework4/design_pattern/cpp/builder.cpp
#include <iostream>
#include <string>

class Person {
public:
    std::string name;
    int age;
    std::string address;

    Person() : name(""), age(0), address("") {}
};

class PersonBuilder {
private:
    Person person;

public:
    PersonBuilder& with_name(const std::string& name) {
        person.name = name;
        return *this; // Return self
    }

    PersonBuilder& with_age(int age) {
        person.age = age;
        return *this; // Return self
    }

    PersonBuilder& with_address(const std::string& address) {
        person.address = address;
        return *this; // Return self
    }

    Person build() {
        return person;
    }
};

int main() {
    Person person = PersonBuilder()
                        .with_name("홍길동")
                        .with_age(20)
                        .with_address("양산시 부산대학로")
                        .build();

    std::cout << "Name: " << person.name << "\n";
    std::cout << "Age: " << person.age << "\n";
    std::cout << "Address: " << person.address << "\n";

    return 0;
}